File: crawler/spiders/weighted_index_spider.py
# encoding: utf-8
# 加權指數
# https://www.twse.com.tw/zh/page/trading/indices/MI_5MINS_HIST.html


import scrapy
from scrapy import signals, Spider
from urllib.parse import urlencode
import time
from random import randint
import datetime
import logging
from copy import copy
from dateutil.relativedelta import relativedelta
import collections
import json

# for cloud function call && scrapy crawl command call
# softlink package folder to root
try:
    from package.tools import check_date, is_settle, format_number
    from package.storage import Storage
except:
    from spiders.package.tools import check_date, is_settle, format_number
    from spiders.package.storage import Storage



class WeightedIndexSpider(scrapy.Spider):
    name = 'weighted_index'
        
    def __init__(self, category=None, *args, **kwargs):
        super(WeightedIndexSpider, self).__init__(*args, **kwargs)
        
        self.dataStorage = Storage('weighted_index')
        self.data = collections.OrderedDict()
        self.today = datetime.date.today()
        self.url = 'https://www.twse.com.tw/indicesReport/MI_5MINS_HIST'
        self.params = {
            'response': 'json',
            'date': '20110101'
        }
        self.startDate = getattr(self, 'start', self.getFormatDate(self.today))
        self.endDate = getattr(self, 'end', self.getFormatDate(self.today))
        self.startObj = self.parseDate(self.startDate)
        self.endObj = self.parseDate(self.endDate)

    def parseDate(self, dateString):
        year = int(dateString[0:4])
        month = int(dateString[4:6])
        day = int(dateString[6:8])
        return {
            'year': year,
            'month': month,
            'day': day,
            'datetime': datetime.date(year, month, day)
        }

    def getFormatDate(self, date):
        year = str(date.year)
        month = str(date.month) if len(str(date.month)) != 1 else "0" + str(date.month)
        day = '01'
        return year + month + day

    # 西元
    def format_ad_date(self, date):
        date_arr = date.split('/')
        year = date_arr[0]
        month = date_arr[1]
        day = date_arr[2]
        year = str(int(year) + 1911)
        return year + '/' + month + '/' + day

    def start_requests(self):
        targetDateObj = copy(self.startObj)
        while(targetDateObj['datetime'] <= self.endObj['datetime']):
            self.params['date'] = self.getFormatDate(targetDateObj['datetime'])
            url = self.url + '?' + urlencode(self.params)
            yield scrapy.Request(
                url=url,
                callback=self.parse,
                cb_kwargs=dict(targetDateObj=copy(targetDateObj)),
                errback=self.handle_failure)

            targetDateObj['datetime'] = targetDateObj['datetime'] + relativedelta(months=1)
            targetDateObj['year'] = targetDateObj['datetime'].year
            targetDateObj['month'] = targetDateObj['datetime'].month

    def handle_failure(self, failure):
        self.log(failure, level=logging.ERROR)
        # try with a new proxy
        self.log('restart from the failed url {}'.format(failure.request.url))
        time.sleep(120)
        yield scrapy.Request(
            url=failure.request.url,
            callback=self.parse,
            cb_kwargs=failure.request.cb_kwargs,
            errback=self.handle_failure)

    def parse(self, response, targetDateObj):
        print(targetDateObj['datetime'])
        result = json.loads(response.text)
        data = result['data']
        year = targetDateObj['year']
        for item in data:
            datestart = self.format_ad_date(item[0])
            if year not in self.data:
                self.data[year] = {}

            self.data[year][datestart] = {}
            self.data[year][datestart]["open"] = format_number(item[1].split(".")[0]) # 開盤
            self.data[year][datestart]["high"] = format_number(item[2].split(".")[0]) # 最高
            self.data[year][datestart]["low"] = format_number(item[3].split(".")[0])  # 最低
            self.data[year][datestart]["w_index"] = format_number(item[4].split(".")[0]) # 收盤
            self.data[year][datestart]["is_settle"] = is_settle(datestart, '/')

    @classmethod
    def from_crawler(cls, crawler, *args, **kwargs):
        spider = super(WeightedIndexSpider, cls).from_crawler(crawler, *args, **kwargs)
        crawler.signals.connect(spider.spider_closed, signal=signals.spider_closed)
        return spider

    def spider_closed(self, spider):
        for year in self.data:
            newData = self.data[year]
            data = dict()

            try:
                data = self.dataStorage.getOldData(year)
            except:
                pass

            data.update(newData)
            self.dataStorage.saveData(year, data)

File: node-crawler/src/crawlers/weighted-index-spider.ts
// 加權指數
// https://www.twse.com.tw/zh/page/trading/indices/MI_5MINS_HIST.html

import { format, addMonths, setDate } from 'date-fns'
import { delay, formatStringNumber, isSettle } from '../utils/index.ts'
import Api from '../api/index.ts'
import {
  YearData,
  DataValue,
  WeightedIndexParams,
  WeightedIndexResElement,
  WeightedIndexRes
} from './weighted-index-spider-types.ts'
import Storage from '../storage/index.ts'
import { saveData, RecordYear } from '../utils/store.ts'

const name = 'weighted_index'
const args: string[] = process.argv.slice(2)

let startDate: Date = args[0]
  ? setDate(new Date(args[0]), 1)
  : setDate(new Date(), 1)

const endDate: Date = args[1]
  ? setDate(new Date(args[1]), 1)
  : setDate(new Date(), 1)

const dataStorage = new Storage(name)

const data: RecordYear<YearData> = {}

// 民國轉西元
function formatAdDate(date: string) {
  const dateArr = date.split('/')
  const year = dateArr[0]
  const month = dateArr[1]
  const day = dateArr[2]
  const adYear = String(parseInt(year) + 1911)
  return adYear + '/' + month + '/' + day
}

function handleResponse(res: WeightedIndexRes) {
  res.forEach((item: WeightedIndexResElement) => {
    const [date, open, high, low, w_index] = item
    const adDate = formatAdDate(date)
    const year = adDate.split('/')[0]

    data[year] = data[year] || {}
    data[year][adDate] = {
      open: formatStringNumber(open),
      high: formatStringNumber(high),
      low: formatStringNumber(low),
      w_index: formatStringNumber(w_index),
      is_settle: isSettle(adDate, '/')
    }
  })
}

async function save() {
  await saveData<DataValue>(data, dataStorage)
}

async function main() {
  console.log(`start - ${name}`)
  while (startDate <= endDate) {
    console.log(`fetching - ${format(startDate, 'yyyyMM01')}`)
    try {
      await delay(2000)
      const params: WeightedIndexParams = {
        response: 'json',
        date: format(startDate, 'yyyyMM01') // 20220101
      }

      const { data } = await Api.getWeightedIndex(params)
      const { stat, data: res } = data

      if (stat !== 'OK' || !data) {
        break
      }

      handleResponse(res)
    } catch (error) {
      console.error(error)
      break
    }
    startDate = addMonths(startDate, 1)
  }

  try {
    await save()
  } catch (error) {
    console.error(error)
  }

  return Promise.resolve()
}

await main()
